getting_started/Data_Input_Pipeline.md

```diff
@@ -15,10 +15,48 @@
+| Grain | ArrayRecord, available through Tensorflow Datasets | fully deterministic, regardless of preemption | only supports random access datasets |
+| TFDS | TFRecord, available through Tensorflow Datasets |  | only supports TFRecords<br>non-deterministic with preemption<br>(deterministic without preemption) |
 
+<!-- TODO (aireenmei): add data on perf -->
+
+### HuggingFace pipeline
+The HuggingFace pipeline supports streaming directly from HuggingFace Hub, or from GCS bucket in HuggingFace supported formats (parquet, json, etc.). This is through the HuggingFace [`datasets.load_dataset` API](https://huggingface.co/docs/datasets/en/loading) with `streaming=True`, which take in `hf_*` parameters. 
+#### Example config for streaming from HuggingFace Hub (no download needed):
+```
+dataset_type: hf
+hf_path: 'allenai/c4'  # for using https://huggingface.co/datasets/allenai/c4
+hf_data_dir: 'en'
+hf_data_files: ''
+# for HF pipeline, tokenizer_path can be a path in HuggingFace Hub, 
+# or a local path containing tokenizer in a format supported by transformers.AutoTokenizer
+tokenizer_path: 'google-t5/t5-large'  # for using https://huggingface.co/google-t5/t5-large
+hf_access_token: ''  # provide token if using gated dataset or tokenizer
+```
+
+#### Example config for streaming from downloaded data in a GCS bucket:
+```
+dataset_type: hf
+hf_path: 'parquet'  # or json, arrow, etc.
+hf_data_dir: ''
+hf_data_files: 'gs://<bucket>/<folder>/*-train-*.parquet'  # match the train files
+# for HF pipeline, tokenizer_path can be a path in HuggingFace Hub, 
+# or a local path containing tokenizer in a format supported by transformers.AutoTokenizer
+tokenizer_path: 'google-t5/t5-large'  # for using https://huggingface.co/google-t5/t5-large
+```
+#### Limitations & Recommendations
+1. Streaming data directly from HuggingFace Hub may be impacted by the traffic of the server. During peak hours you may encounter "504 Server Error: Gateway Time-out". It's recommended to download the HuggingFace dataset to a GCS buckt or disk for the most stable experience.
+2. Streaming data directly from HuggingFace Hub works in multihost settings with a samll number of hosts. We have encountered "read time out" error with host number > 16.
+3. Multihost data input is achieved through HuggingFace's `datasets.distributed.split_dataset_by_node` API, which is more performant when (data shards/files) % (number of host reading data) == 0. To give MaxText users a performant experience, we follow this rule and will use data shards up to the number that's divisible by number of host reading data. For instance if your dataset has 100 shards, and you use 8 hosts, at most 96/100 shards of data will be used since 96%8==0. You can reshard or combine shards if you want to utilize the whole dataset.
+
+### Grain pipeline - for determinism
+
+<!-- TODO (aireenmei): add more details/examples on why determinism is important -->
```

Comment:
One example where deterministic replay is crucial is sensitive convergence experiments such as testing quantization techniques. When comparing a quantized vs unquantized run we want everything else to be identical, in particular the data in each batch. Such experiments may be long and require saving/resuming, and we want the data order to be replyable even if the paired runs save/resume at different steps. Another use case is to debug training spikes/anomalies - if we can replay the exact data it can help debug a bad data batch versus a hardware/SDC issue.

Reply:
Thanks for the examples! I added a few more paragraphs there and incoporated your examples, with Gemimi's help :)